spec/lib/diaspora_federation/discovery/discovery_spec.rb

```diff
@@ -88,66 +91,192 @@ module DiasporaFederation
+              .to_return(status: 200, body: host_meta_xrd)
+            stub_request(:get, "http://localhost:3000/.well-known/webfinger.xml?resource=acct:#{account}")
+              .to_return(status: 200, body: webfinger_xrd)
+            stub_request(:get, "http://localhost:3000/hcard/users/#{alice.guid}")
+              .to_return(status: 200, body: hcard_html)
+
+            expect_callback(:save_person_after_webfinger, kind_of(Entities::Person))
+            person = subject.fetch_and_save
+
+            expect(person.guid).to eq(alice.guid)
+            expect(person.diaspora_id).to eq(account)
+          end
+        end
+
+        context "http fallback enabled" do
+          before do
+            DiasporaFederation.webfinger_http_fallback = true
+          end
+
+          it "falls back to http if https fails with 404" do
+            stub_request(:get, "https://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
+              .to_return(status: 404)
+            stub_request(:get, "http://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
+              .to_return(status: 200, body: webfinger_jrd)
+            stub_request(:get, "http://localhost:3000/hcard/users/#{alice.guid}")
+              .to_return(status: 200, body: hcard_html)
+
+            expect_callback(:save_person_after_webfinger, kind_of(Entities::Person))
+            person = subject.fetch_and_save
+
+            expect(person.guid).to eq(alice.guid)
+            expect(person.diaspora_id).to eq(account)
+          end
+
+          it "falls back to http if https fails with ssl error" do
+            stub_request(:get, "https://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
+              .to_raise(OpenSSL::SSL::SSLError)
+            stub_request(:get, "http://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
+              .to_return(status: 200, body: webfinger_jrd)
+            stub_request(:get, "http://localhost:3000/hcard/users/#{alice.guid}")
+              .to_return(status: 200, body: hcard_html)
+
+            expect_callback(:save_person_after_webfinger, kind_of(Entities::Person))
+            person = subject.fetch_and_save
+
+            expect(person.guid).to eq(alice.guid)
+            expect(person.diaspora_id).to eq(account)
+          end
+
+          after do
+            DiasporaFederation.webfinger_http_fallback = false
+          end
+        end
       end
 
-      it "fails if the diaspora* ID was not found" do
-        stub_request(:get, "https://localhost:3000/.well-known/host-meta")
-          .to_return(status: 200, body: host_meta_xrd)
-        stub_request(:get, "http://localhost:3000/.well-known/webfinger.xml?resource=acct:#{account}")
-          .to_return(status: 404)
+      context "legacy WebFinger" do
+        it "falls back to legacy WebFinger" do
+          stub_request(:get, "https://localhost:3000/.well-known/webfinger?resource=acct:#{account}")
```

Comment:
No, I wanted to test the different scenarios here.

And older pods (since 0.6.0.0) did have RFC 7033 WebFinger (for OpenID), just without any usable data, so the "real world scenario" would be a `ValidationError` because of the missing data, not a 404. So I changed the `"foobar"` (which raised a `JSON::ParserError`) to the JSON an old pod will return, to simulate the most common scenario.

But I still want to test `404` and `SSLError` here.